src/traces/carpet/axis_defaults.js

```diff
@@ -0,0 +1,238 @@
+
+    if(autoRange) coerce('rangemode');
+
+    coerce('range');
+    containerOut.cleanRange();
+
+    coerce('fixedrange');
+
+    handleTickValueDefaults(containerIn, containerOut, coerce, axType);
+    handleTickLabelDefaults(containerIn, containerOut, coerce, axType, options);
+    handleCategoryOrderDefaults(containerIn, containerOut, coerce);
+
+    var gridColor = coerce2('gridcolor', addOpacity(dfltColor, 0.3));
+    var gridWidth = coerce2('gridwidth');
+    var showGrid = coerce('showgrid');
+
+    if(!showGrid) {
+        delete containerOut.gridcolor;
+        delete containerOut.gridwidth;
+    }
+
+    var startLineColor = coerce2('startlinecolor', dfltColor);
+    var startLineWidth = coerce2('startlinewidth', gridWidth);
+    var showStartLine = coerce('startline', containerOut.showgrid || !!startLineColor || !!startLineWidth);
+
+    if(!showStartLine) {
+        delete containerOut.startlinecolor;
+        delete containerOut.startlinewidth;
+    }
+
+    var endLineColor = coerce2('endlinecolor', dfltColor);
+    var endLineWidth = coerce2('endlinewidth', gridWidth);
+    var showEndLine = coerce('endline', containerOut.showgrid || !!endLineColor || !!endLineWidth);
+
+    if(!showEndLine) {
+        delete containerOut.endlinecolor;
+        delete containerOut.endlinewidth;
+    }
+
+    if(!showGrid) {
+        delete containerOut.gridcolor;
+        delete containerOut.gridWidth;
+    } else {
+        coerce('minorgridcount');
+        coerce('minorgridwidth', gridWidth);
+        coerce('minorgridcolor', addOpacity(gridColor, 0.06));
+
+        if(!containerOut.minorgridcount) {
+            delete containerOut.minorgridwidth;
+            delete containerOut.minorgridcolor;
+        }
+    }
+
+    containerOut._separators = options.fullLayout.separators;
+
+    // fill in categories
+    containerOut._initialCategories = axType === 'category' ?
+        orderedCategories(letter, containerOut.categoryorder, containerOut.categoryarray, options.data) :
+        [];
+
+    if(containerOut.showticklabels === 'none') {
+        delete containerOut.tickfont;
+        delete containerOut.tickangle;
+        delete containerOut.showexponent;
+        delete containerOut.exponentformat;
+        delete containerOut.tickformat;
+        delete containerOut.showticksuffix;
+        delete containerOut.showtickprefix;
+    }
+
+    if(!containerOut.showticksuffix) {
+        delete containerOut.ticksuffix;
+    }
+
+    if(!containerOut.showtickprefix) {
+        delete containerOut.tickprefix;
+    }
+
+    // It needs to be coerced, then something above overrides this deep in the axis code,
+    // but no, we *actually* want to coerce this.
+    coerce('tickmode');
+
+    if(!containerOut.title || (containerOut.title && containerOut.title.length === 0)) {
+        delete containerOut.titlefont;
+        delete containerOut.titleoffset;
+    }
+
+    return containerOut;
+};
+
+function setAutoType(ax, data) {
```

Comment:
I think *maybe* there was a conflict, but I merge with master and sorted out a couple trivial merge conflicts and verified that axis constraints work.